File: proposed-projects/round-trip-experiment/reverse/20260909T180307Z_280f3890/20260910T095157Z_2fde87b4/ported_implementation/gbnf/utils/errors/build_error_position.py
from typing import List, Optional

MAXIMUM_NUMBER_OF_ERROR_LINES_TO_SHOW = 3


def build_error_position(src: str, pos: int) -> List[str]:
    if src == "":
        return ["No input provided"]
    lines = src.split("\n")

    line_idx = 0
    line: Optional[str] = lines[line_idx] if line_idx < len(lines) else None
    while line and pos > len(line) - 1 and pos < len(src):
        pos -= len(line)
        line_idx += 1
        line = lines[line_idx] if line_idx < len(lines) else None

    lines_to_show = lines[
        max(0, line_idx - (MAXIMUM_NUMBER_OF_ERROR_LINES_TO_SHOW - 1)) : line_idx + 1
    ]

    return [*lines_to_show, f"{' ' * max(0, pos)}^"]

File: proposed-projects/round-trip-experiment/reverse/20260909T154334Z_609edff1/20260910T091410Z_185eaa51/ported_implementation/src/utils/errors/build-error-position.ts
import { itemAt } from '../char-at';

export const MAXIMUM_NUMBER_OF_ERROR_LINES_TO_SHOW = 3;

export const buildErrorPosition = (src: string, pos: number): string[] => {
  if (src === '') {
    return ['No input provided'];
  }

  const lines = src.split('\n');

  let lineIdx = 0;
  while (itemAt(lines, lineIdx) && pos > lines[lineIdx].length - 1) {
    pos -= lines[lineIdx].length;
    lineIdx += 1;
  }

  const linesToShow: string[] = [];
  const start = Math.max(0, lineIdx - (MAXIMUM_NUMBER_OF_ERROR_LINES_TO_SHOW - 1));
  for (let i = start; i <= lineIdx; i++) {
    const line = itemAt(lines, i);
    linesToShow.push(line === undefined ? '' : line);
  }
  return [
    ...linesToShow,
    `${' '.repeat(Math.max(0, pos))}^`,
  ];
};
